File: Java/best-time-to-buy-and-sell-stock.java
//  Best Time to Buy and Sell Stock

// Time O(N), Space O(1)
class Solution {
    public int maxProfit(int[] prices) {
        if (prices.length < 2)
            return 0;
        int max_profit = 0, min_price = prices[0];
        for (int i=1; i<prices.length; i++) {
            max_profit = Math.max(max_profit, prices[i] - min_price);
            min_price = Math.min(min_price, prices[i]);
        }
        return max_profit;
    }
}

File: Leetcode150/best-time-to-buy-and-sell-stock.cc
//  Best Time to Buy and Sell Stock

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() < 2)  return 0;
        int max_profit = 0, min_price = prices[0];
        int buy_day = 0, sell_day = -1, pre_buy_day = 0;
        for (int i=1; i<prices.size(); i++) {
            if (prices[i] - min_price > max_profit) {
                max_profit = prices[i] - min_price;
                sell_day = i;
            }
            if (prices[i] < min_price) {
                pre_buy_day = i;
                min_price = prices[i];
            }
        }
        cout << pre_buy_day << "\t" << sell_day << endl;
        return max_profit;
    }
};